src/journal/bsod.c

```diff
@@ -0,0 +1,209 @@
+        sd_journal *j = NULL;
+        const void *d;
+        size_t l;
+        sd_id128_t boot_id;
+        char boot_id_filter[42];
+        int r;
+
+        r = sd_journal_open(&j, SD_JOURNAL_LOCAL_ONLY);
+        if (r < 0) {
+                log_error_errno(r, "Failed to open journal: %m");
+                goto clean_journal;
+        }
+
+        r = sd_id128_get_boot(&boot_id);
+        if (r < 0) {
+                log_error_errno(r, "Failed to get boot ID: %m");
+                goto clean_journal;
+        }
+
+        snprintf(boot_id_filter, sizeof(boot_id_filter), "_BOOT_ID=" SD_ID128_FORMAT_STR, SD_ID128_FORMAT_VAL(boot_id));
+
+        r = sd_journal_add_match(j, boot_id_filter, 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add boot ID filter: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_add_match(j, "_UID=0", 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add User ID filter: %m");
+                goto clean_journal;
+        }
+
+        assert_cc(0 == LOG_EMERG);
+        r = sd_journal_add_match(j, "PRIORITY=0", 0);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to add Emergency filter: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_seek_head(j);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to seek to start of jornal: %m");
+                goto clean_journal;
+        }
+
+        r = sd_journal_next(j);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to read next journal entry: %m");
+                goto clean_journal;
+        }
+
+        if (r == 0) {
+                log_warning("No entries in the journal");
+                r = -ENOENT;
+                goto clean_journal;
+        }
+
+        r = sd_journal_get_data(j, "MESSAGE", &d, &l);
+        if (r < 0) {
+                log_warning_errno(r, "Failed to read journal message: %m");
+                goto clean_journal;
+        }
+        *ret = strndup((char *) d, l);
+        if (! *ret) {
+                r = -ENOMEM;
+                log_warning_errno(r, "Failed to duplicate journal message: %m");
+                goto clean_journal;
+        }
+
+clean_journal:
+        sd_journal_close(j);
+        return r;
+}
+
+static int display_emergency_message_fullscreen(char * message) {
+        assert(message);
+        int r;
+        size_t i;
+        struct vt_stat terminal_status;
+        char tty[12];
+        _cleanup_close_ int fd = -EBADF;
+
+        fd = open_terminal("/dev/tty1", O_RDWR|O_NOCTTY|O_CLOEXEC);
+        if (fd < 0)
+                return -errno;
+
+        r = ioctl(fd, VT_GETSTATE, &terminal_status);
+        if (r < 0)
+                return -errno;
+
+        for (i = 0; i < sizeof(terminal_status.v_state) * 8; i++) {
+                if ((terminal_status.v_state & (1 << i)) == 0)
+                        break;
+        }
```

Comment:
maybe turn the ioctl and the for loop into a function of its own `int find_next_free_vt(int fd)`

Reply:
as @halfline suggested, please rework this to use VT_OPENQRY